### PR TITLE
BUG-119682:Handling registration of multiple Clusters

### PR DESCRIPTION
**Pull request by PankajWorks**
Tested scenarios
## one Cluster managed by CM

- register the cluster

```bash
Tell me about Cloudera Manager Instance
CM URL (http(s)://host:[port]) [http://abc:port]:
CM admin username [admin]:
CM admin password:
Checking communication between DataPlane and Cloudera Manager ...
Total clusters managed by Cloudera Manager Instance : 1
Registering cluster to DataPlane...
Enter details for cluster : Cluster 1
Data Center Name [DC1]: DC2
Cluster Descriptions [test]:
Cluster : Cluster 1 is registered with id : 44
Success! You are all set, your cluster is registered and ready to use.
```

## Two Cluster managed by CM 

-  Register ALL
```bash
Tell me about Cloudera Manager Instance
CM URL (http(s)://host:[port]) [http://abc:port]: 
CM admin username [admin]: 
CM admin password: 
Checking communication between DataPlane and Cloudera Manager ...
Total clusters managed by Cloudera Manager Instance : 2
Clusters already registered in DataPlane : 
Clusters which can be registered in DataPlane : Cluster 1,Cluster 2
Register all y/n [y]: y

Clusters which will be registered in DataPlane : Cluster 1,Cluster 2
Registering cluster to DataPlane...
Enter details for cluster : Cluster 1
Data Center Name [DC1]: 
Cluster Descriptions [test]: 
Enter details for cluster : Cluster 2
Data Center Name [DC1]: 
Cluster Descriptions [test]: 
Cluster : Cluster 1 is registered with id : 40 
Cluster : Cluster 2 is registered with id : 39 
Success! You are all set, your cluster is registered and ready to use.
```

- Register one cluster from file input

```bash
Tell me about Cloudera Manager Instance
CM URL (http(s)://host:[port]) [http://abc:port]: 
CM admin username [admin]: 
CM admin password: 
Checking communication between DataPlane and Cloudera Manager ...
Total clusters managed by Cloudera Manager Instance : 2
Clusters already registered in DataPlane : 
Clusters which can be registered in DataPlane : Cluster 1,Cluster 2
Register all y/n [y]: n
Enter full path of file containing cluster names  [cluster.txt]: 

Clusters which will be registered in DataPlane : Cluster 1
Registering cluster to DataPlane...
Enter details for cluster : Cluster 1
Data Center Name [DC1]: 
Cluster Descriptions [test]: 
Cluster : Cluster 1 is registered with id : 41
``` 

- Register remaining
```bash
Tell me about Cloudera Manager Instance
CM URL (http(s)://host:[port]) [http://abc:port]: 
CM admin username [admin]: 
CM admin password: 
Checking communication between DataPlane and Cloudera Manager ...
Total clusters managed by Cloudera Manager Instance : 2
Clusters already registered in DataPlane : Cluster 1
Clusters which can be registered in DataPlane : Cluster 2
Register all y/n [n]: y

Clusters which will be registered in DataPlane : Cluster 2
Registering cluster to DataPlane...
Enter details for cluster : Cluster 2
Data Center Name [DC1]: 
Cluster Descriptions [test]: 
Cluster : Cluster 2 is registered with id : 42 
Success! You are all set, your cluster is registered and ready to use.
```

### CODE REVIEW

File: dp-cluster-setup-utility.py [hunk 1]
```diff
@@ -1016,25 +1028,31 @@ def additional_request_for_dp_1_3_and_above(self,ambari, knox):
...
-      'behindGateway': False,
-      'knoxEnabled': False,
-      'managerType': "cloudera-manager",
-      'clusterType': cm.cluster.type,
-      'properties': {'tags': []}
-    }]
+  def registration_request_cm(self, cm, user, cluster_names):
+    registration_request = []
+    for cluster_name in cluster_names:
+      cluster_obj = next(obj for obj in cm.clusters if obj.cluster_name == cluster_name)
```
Comment (yhemanth) [inline dp-cluster-setup-utility.py]: This code (or something similar) seems simpler to understand:
```cluster_obj = filter(lambda c: c.cluster_name == cluster_name, cm.clusters)[0]```

File: dp-cluster-setup-utility.py [hunk 2]
```diff
@@ -1568,24 +1585,24 @@ def security_type_supported(self):
...
+  def stack_supported(self, cluster):
+    stack = cluster.installed_stack()
     check_version = False
     (major,minor) = stack.version.split('.')[:2]
-    if (major == '5' and int(minor) >= 17) or (major == '6' and int(minor) >= 1):
+    if (major == '5' and int(minor) >= 17) or (major == '6' and int(minor) >= 3):
       check_version = True
     return stack.name == 'CDH' and check_version
 
   def security_type_supported(self):
```
Comment (yhemanth) [inline dp-cluster-setup-utility.py]: Can remove this for now.

Comment (PankajWorks) [reply dp-cluster-setup-utility.py]: Should we remove both major and minor version check or only minor version?

File: dp-cluster-setup-utility.py [hunk 3]
```diff
@@ -1743,24 +1761,55 @@ def __init__(self):
...
+    elif cm.total_clusters > 1:
+      print "Clusters already registered in DataPlane : %s" % ','.join([cluster for cluster in clusters_registered])
+      print "Clusters which can be registered in DataPlane : %s" % ','.join([cluster for cluster in clusters_not_registered])
+      if len(clusters_not_registered) > 0:
+        install_all = user.decision('%s y/n' % "Register all", "cm.register_all", default=False)
+        if install_all:
+          clusters_to_register = clusters_not_registered
+        else:
+          user_provided_clusters = []
+          cluster_input_file = user.input('Enter full path of file containing cluster names ', 'cm.cluster_file')
```
Comment (yhemanth) [inline dp-cluster-setup-utility.py]: We will need some more informative text for this flow. Can improve after we this patch in.

File: dp-cluster-setup-utility.py [hunk 4]
```diff
@@ -614,34 +614,36 @@ def __init__(self, base_url, credentials=Credentials('admin', 'admin'), api_vers
...
+
+  def _find_clusters(self):
     try:
       response = self.client.cluster_api_instance().read_clusters(view='full')
-      return response.items[0]
+      return response.items
     except ApiException as e:
       raise NoClusterFound(e)
 
+  def get_cluster_instance(self, cluster_name):
```
Comment (yhemanth) [inline dp-cluster-setup-utility.py]: This doesn't seem to be used. Can remove it.

File: dp-cluster-setup-utility.py [hunk 5]
```diff
@@ -1568,24 +1586,24 @@ def security_type_supported(self):
...
     return True
 
-  def stack_supported(self):
-    stack = self.cm.installed_stack()
+  def stack_supported(self, cluster):
+    stack = cluster.installed_stack()
     check_version = False
     (major,minor) = stack.version.split('.')[:2]
-    if (major == '5' and int(minor) >= 17) or (major == '6' and int(minor) >= 1):
+    if (major == '5' and int(minor) >= 17):
```
Comment (yhemanth) [inline dp-cluster-setup-utility.py]: Why did this change?

**Pull request merged**